File: Java/java9/sort/stringSort/src/main/java/packt/java9/by/example/ch03/bubble/BubbleSort.java
package packt.java9.by.example.ch03.bubble;

import packt.java9.by.example.ch03.*;
import java.util.Comparator;

public class BubbleSort implements Sort {
    @Override
    public void sort(SortableCollection collection) {
        int n = collection.size();
        while (n > 1) {
            for (int j = 0; j < n - 1; j++) {
                if (comparator.compare(collection.get(j),
                            collection.get(j + 1)) > 0) {
                    swapper.swap(j, j + 1);
                }
            }
        }
    }

    private Comparator comparator = null;

    @Override
    public void setComparator(Comparator comparator) {
        this.comparator = comparator;
    }

    private Swapper swapper = null;

    @Override
    public void setSwapper(Swapper swapper) {
        this.swapper = swapper;
    }
}

File: algorithms/js/algorithms/bubble-sort.js
const { checkArray, swap } = require('../util/arrayBasics');

function bubble(array) {
  checkArray(array);
  const newArray = array;

  for (let i = array.length - 1; i > 0; i--) {
    for (let j = 0; j < i; j++) {
      if (array[j] > array[j+1]) swap(array, j, j+1);
    }
  }

  return newArray;
}

console.log(bubble([1,44, 3, 72, 9]));
